filters/include/pcl/filters/experimental/voxel_grid.h

```diff
@@ -0,0 +1,617 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2021-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/common/centroid.h>
+#include <pcl/common/common.h>
+#include <pcl/filters/experimental/cartesian_filter.h>
+#include <pcl/filters/filter.h>
+
+#include <unordered_map>
+
+namespace pcl {
+namespace experimental {
+
+template <typename PointT>
+struct Voxel {
+
+  union Centroid {
+    Centroid() {}
+    ~Centroid() {}
+
+    CentroidPoint<PointT> all_fields;
+    Eigen::Array4f xyz;
+  };
```

Comment:
Do you think some alignment issue due to Eigen is causing the issue here? CentroidPoint uses `PCL_MAKE_ALIGNED_OPERATOR_NEW` and Eigen definitely has it's own alignment.

Could we try using something similar?

Reply:
I tried both macro from PCL and Eigen and it's still the same.

Reply:
Cpuld you try to use valgrind to find why there's a weird issue with stack data and not heap data?

Reply:
I'm still investigating this but the trace always go deep into stl library makes it difficult to debug

FYI the track by valgrind
```rust
==120237== Conditional jump or move depends on uninitialised value(s)
==120237==    at 0x125F59: std::_Rb_tree<unsigned int, std::pair<unsigned int const, unsigned long>, std::_Select1st<std::pair<unsigned int const, unsigned long> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >::_M_erase(std::_Rb_tree_node<std::pair<unsigned int const, unsigned long> >*) (stl_tree.h:1911)
==120237==    by 0x1232FB: std::_Rb_tree<unsigned int, std::pair<unsigned int const, unsigned long>, std::_Select1st<std::pair<unsigned int const, unsigned long> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >::clear() (stl_tree.h:1266)
==120237==    by 0x12AB8B: std::_Rb_tree<unsigned int, std::pair<unsigned int const, unsigned long>, std::_Select1st<std::pair<unsigned int const, unsigned long> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >::_M_move_assign(std::_Rb_tree<unsigned int, std::pair<unsigned int const, unsigned long>, std::_Select1st<std::pair<unsigned int const, unsigned long> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >&, std::integral_constant<bool, true>) (stl_tree.h:1681)
==120237==    by 0x125E65: std::_Rb_tree<unsigned int, std::pair<unsigned int const, unsigned long>, std::_Select1st<std::pair<unsigned int const, unsigned long> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >::operator=(std::_Rb_tree<unsigned int, std::pair<unsigned int const, unsigned long>, std::_Select1st<std::pair<unsigned int const, unsigned long> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >&&) (stl_tree.h:1723)
==120237==    by 0x120FD4: std::map<unsigned int, unsigned long, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >::operator=(std::map<unsigned int, unsigned long, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, unsigned long> > >&&) (stl_map.h:100)
==120237==    by 0x121002: pcl::detail::AccumulatorLabel::operator=(pcl::detail::AccumulatorLabel&&) (accumulators.hpp:211)
==120237==    by 0x121030: boost::fusion::vector_detail::store<2ul, pcl::detail::AccumulatorLabel>::operator=(boost::fusion::vector_detail::store<2ul, pcl::detail::AccumulatorLabel>&&) (vector.hpp:156)
==120237==    by 0x121094: boost::fusion::vector_detail::vector_data<std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, pcl::detail::AccumulatorXYZ, pcl::detail::AccumulatorRGBA, pcl::detail::AccumulatorLabel>::operator=(boost::fusion::vector_detail::vector_data<std::integer_sequence<unsigned long, 0ul, 1ul, 2ul>, pcl::detail::AccumulatorXYZ, pcl::detail::AccumulatorRGBA, pcl::detail::AccumulatorLabel>&&) (vector.hpp:184)
==120237==    by 0x1210C2: boost::fusion::vector<pcl::detail::AccumulatorXYZ, pcl::detail::AccumulatorRGBA, pcl::detail::AccumulatorLabel>::operator=(boost::fusion::vector<pcl::detail::AccumulatorXYZ, pcl::detail::AccumulatorRGBA, pcl::detail::AccumulatorLabel>&&) (vector.hpp:262)
==120237==    by 0x121106: pcl::CentroidPoint<pcl::PointXYZRGBL>::operator=(pcl::CentroidPoint<pcl::PointXYZRGBL>&&) (centroid.h:1022)
==120237==    by 0x1211B5: pcl::experimental::Voxel<pcl::PointXYZRGBL>::Voxel(bool) (voxel_grid.h:37)
==120237==    by 0x11D7D8: pcl::experimental::LabeledVoxel::LabeledVoxel(bool) (voxel_grid_label.h:27)
```

It goes through several move assign, and when I tried to declare my move assign function the trace go somewhere also related to `std::map` in `AccumulatorLabel`

I'm trying to use `boost::variant` since it is only a bit slower

Reply:
Since we will transit to `std::variant` anyway once we move to C++17, I think it is better to use `boost::variant` so that the transition can be easier. 

althrough we might sacrifice some performance in theory (`boost::get` has a runtime check)

## Performance
But runtime results are very similar in my benchmark

### union
```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
table_VoxelGrid                         6.26 ms         6.26 ms          111
table_ApproximateVoxelGrid              5.08 ms         5.08 ms          137
table_FunctorVoxelGrid                  3.99 ms         3.99 ms          174
milk_cartoon_VoxelGrid                  8.01 ms         8.01 ms           88
milk_cartoon_ApproximateVoxelGrid       5.49 ms         5.49 ms          126
milk_cartoon_FunctorVoxelGrid           4.61 ms         4.61 ms          151
office_VoxelGrid                        13.4 ms         13.4 ms           51
office_ApproximateVoxelGrid             8.10 ms         8.10 ms           85
office_FunctorVoxelGrid                 11.7 ms         11.7 ms           55
five_people_VoxelGrid                   14.0 ms         14.0 ms           51
five_people_ApproximateVoxelGrid        7.48 ms         7.47 ms           92
five_people_FunctorVoxelGrid            8.50 ms         8.50 ms           76
```

### boost::variant

```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
table_VoxelGrid                         6.42 ms         6.42 ms          110
table_ApproximateVoxelGrid              5.09 ms         5.08 ms          138
table_FunctorVoxelGrid                  4.16 ms         4.16 ms          168
milk_cartoon_VoxelGrid                  8.17 ms         8.17 ms           85
milk_cartoon_ApproximateVoxelGrid       5.56 ms         5.56 ms          122
milk_cartoon_FunctorVoxelGrid           4.87 ms         4.87 ms          139
office_VoxelGrid                        13.6 ms         13.6 ms           52
office_ApproximateVoxelGrid             8.08 ms         8.08 ms           85
office_FunctorVoxelGrid                 11.3 ms         11.3 ms           59
five_people_VoxelGrid                   13.7 ms         13.7 ms           51
five_people_ApproximateVoxelGrid        7.49 ms         7.49 ms           92
five_people_FunctorVoxelGrid            8.38 ms         8.38 ms           80
```